TRD for UART HIL (#2756)
tock/tock: doc/reference/trd-uart.md

    fn get_width(&self) -> Width;
    fn get_parity(&self) -> Parity;
    fn get_stop_bits(&self) -> StopBits;
    fn get_flow_control(&self) -> bool;
    fn get_configuration(&self) -> Configuration;
}

pub trait Configure {
    fn set_baud_rate(&self, rate: u32) -> Result<u32, ErrorCode>;
    fn set_width(&self, width: Width) -> Result<(), ErrorCode>;
    fn set_parity(&self, parity: Parity) -> Result<(), ErrorCode>;
    fn set_stop_bits(&self, stop: StopBits) -> Result<(), ErrorCode>;
    fn set_flow_control(&self, on: bool) -> Result<(), ErrorCode>;
    fn configure(&self, params: Parameters) -> Result<(), ErrorCode>;
}
```

Methods in `Configure` can return the following error conditions:
  - `OFF`: The underlying hardware is currently not available, perhaps
    because it has not been initialized or in the case of a shared
    hardware USART controller because it is set up for SPI.
  - `INVAL`: Baud rate was set to 0.
  - `ENOSUPPORT`: The underlying UART cannot satisfy this configuration.
  - `FAIL`: Other failure condition.

The UART may be unable to set the precise baud rate specified. For
example, the UART may be driven off a fixed clock with integer

> prescalar. A call to `configure` MUST set the baud rate to the closest
> possible value to the `baud_rate` field of the `params` argument and a
> call to `set_baud_rate` MUST set the baud rate to the closest possible
> value to the `rate` argument. The `Ok` result of `set_baud_rate`
> includes the actual rate set, while an `Err(INVAL)` result means the
> requested rate is well outside the operating speed of the UART (e.g., 4MHz).

Reviewer: ```suggestion
prescalar. In response to a call to `configure`, the underlying UART MUST set the baud rate to the closest
possible value to that requested by the `baud_rate` field of the `params` argument, and the response to a
call to `set_baud_rate` MUST set the baud rate to the closest possible
value to the provided `rate` argument. The `Ok` result of `set_baud_rate`
includes the actual rate set, while an `Err(INVAL)` result means the
requested rate is well outside the operating speed of the UART (e.g., 4MHz).
```

I eventually got it, but I had to read this many times as my first parse kept interpreting this as the _caller_ need to set the baud rate correctly when calling; this is a bit more verbose but hopefully clearer to other thick-headed readers such as myself 😅

Author: Yeah, I get the confusion. But "in response" seems to suggest asynchrony. How about "Calling `configure` MUST ..."

Author: Yeah, I get the confusion. But "in response" seems to suggest asynchrony. How about "Calling `configure` MUST ..."

3 `Transmit` and `TransmitClient`
===============================

The `Transmit` and `TransmitClient` traits allow a client to transmit
bytes over the UART.

```rust
enum AbortResult {
    Failure,
    Success,

> }

Reviewer: What is the purpose of including this enum instead of of just returning `Ok` or `Err(FAIL)`?

Author: The `Ok` vs. `Err` is about whether there will be a callback. If the return type is `Ok`, there will be a callback. If it's `Err`, there won't be (e.g., because no operation was oustanding). The `AbortResult` type is about whether the operation is actually aborted. So:
  - `Ok(AbortResult::Failure)` means there will be a callback and the operation wasn't aborted
  - `Ok(AbortResult::Success)` means there will be a callback and the operation was aborted
  - `Err(())` means there will not be a callback.

This trinary result really doesn't fit that well with `Result`, but I figured it was better to keep with linguistic support than roll a new type.

Reviewer 2: Given that the Error in this case is just the unit type, this seems better suited to return `Option<AbortResult>`, or to just a custom type like you said. The `?` operator makes it easy to bubble errors up the call stack, but in this case we don't want callers to bubble up the error immediately, because one level higher than the caller it will not be clear that the `Err` does not indicate an actual error, but merely that there will not be a callback.

Reviewer: Okay, I have read all of this now, and I think I have convinced myself that making abort split-phase is more confusing than helpful. The configure trait is already all synchronous, so we have implicitly discarded the possibility of an off-chip implementation requiring everything to be split. The HIL guidelines push towards split, but acknowledge pragmatism for synchronous calls.

Here, I put myself in the userspace position—cancelling operations is hard enough, but now I have to handle two separate asynchronous callbacks (the abort response and the original xmit response) in unknown order? What other case do we have multiple intertwined callbacks in flight for one driver?

Author: > I think what I am missing is why we need to generate a `transmit_abort` callback in the case of `AbortResult::Failure` -- there will still be the regular `transmitted_[word,buffer]` callback. Or perhaps equivalently, when `Err(())` would ever be generated in lieu of an `Ok(AbortResult::Failure)` (nothing to cancel?).
> 
> I'm also concerned about whether we have to enforce any ordering between the `abort` callback and the `transmitted_X` callback.
> 
> I think this is explained in more depth in the text below I didn't get to on first pass; will read that then circle back here.

There is no abort callback -- the callback it's talking about is the transmitted_{word,buffer} callback.

Author: > Given that the Error in this case is just the unit type, this seems better suited to return `Option<AbortResult>`, or to just a custom type like you said. The `?` operator makes it easy to bubble errors up the call stack, but in this case we don't want callers to bubble up the error immediately, because one level higher than the caller it will not be clear that the `Err` does not indicate an actual error, but merely that there will not be a callback.

The reasoning was less about bubbling up and more about trying to make it consistent with the transmit calls (`Ok` means callback, `Err` means no callback). But I agree that the benefits of this consistency are pretty tenuous. You are unlikely to have similar state transition code structures, because the call to cancel doesn't have a buffer it has to deal with in response to `Err`. 

I prefer the `Option` approach, mostly because I think the big deal is callback or no callback, not whether the operation was cancelled (you'll find that out later). So it's nice to be able to match on `Some`/`None` in the base case. Or it could be a type like

```rust 
enum AbortResult {
  Callback(bool),
  NoCallback,
}```

pub trait Transmit<'a> {
    fn set_transmit_client(&self, client: &'a dyn TransmitClient);

    fn transmit_buffer(
        &self,
        tx_buffer: &'static mut [u8],
        tx_len: usize,
    ) -> Result<(), (ErrorCode, &'static mut [u8])>;

>     fn transmit_word(&self, word: u32) -> Result<(), ErrorCode>;

Reviewer: Should this 'no-buffer' path be `word` or `byte`? I can't think of a ton of scenarios where a caller wants to send exactly 4 bytes over UART?

Author: It's a word in case the UART is 9-bit data.

    fn transmitted_buffer(
        &self,
        tx_buffer: &'static mut [u8],
        tx_len: usize,
        rval: Result<(), ErrorCode>,
    );
}
```

The `Transmit` trait has two data paths: `transmit_word` and
`transmit_buffer`.  The `transmit_word` method is used in narrow use
cases where the cost and complexity of buffer management is not
needed. Generally, software should use the `transmit_buffer`
method. Most software implementations use DMA, such that a call to
`transmit_buffer` triggers a single interrupt when the transfer
completes; this saves energy and CPU cycles over per-byte transfers
and also improves transfer speeds because hardware can keep the UART
busy.

Each byte transmitted is a data word for the UART. If the UART is
using 8-bit data words, each data word is a byte. If the UART is using
smaller data words, it MUST ignore the high order bits of the data
values. For example, if the UART is using 6-bit data words and is told
to transmit `0xff`, it will transmit `0x3f`, ignoring the first two
bits. If a client needs to transmit data words larger than 8 bits, it
should use `transmit_word`, as `transmit_buffer` is a buffer of 8-bit

> bytes and cannot store 9-bit values.

Reviewer: Is this case even possible? The `Width` enum above only goes to 8 bits. I suspect 9-bit UART to be such a rarity in modern hardware perhaps that we should omit it at this stage, and then re-evaluate later what the best approach is (i.e. one would still hope to use DMA; perhaps the engine expects 16-bit halfword streams of 9-bit data? I think the 9-bit interface is a question best answered by hardware that uses it)

Author: https://community.silabs.com/s/article/uart-using-9-bit-mode-of-the-uart?language=en_US

There can be all kinds of weird hardware you want to connect to: I'd hate for someone who has to talk 9N1 (on a platform that can) having to start hacking peripheral drivers to do so... The fact that it's only word-level means that there isn't great support, but at least there's *something*.

`Width` should probably include 9?

There can be a single transmit operation ongoing at any
time. Successfully calling either `transmit_buffer` or
`transmit_word` causes the UART to become busy until it issues the
callback corresponding to the outstanding operation.

3.1 `transmit_buffer` and `transmitted_buffer`
===============================

`Transmit::transmit_buffer` sends a buffer of data. The result
returned by `transmit_buffer` indicates whether there will be a
callback in the future. If `transmit_buffer` returns `Ok(())`,
implementation MUST call the `TransmitClient::transmitted_buffer`
callback in the future when the transmission completes or fails. If
`transmit_buffer` returns `Err` it MUST NOT issue a callback in the
future in response to this call. If the error is `BUSY`, this is

> because there is an outstanding call to `transmit_buffer` or
> `transmit_word`: the implementation handles these calls normally and
> issues a callback for them. However, it does not issue a callback for
> the call to `transmit_buffer` that returned `Err`.

Reviewer: ```suggestion
because there is an outstanding call to `transmit_buffer` or
`transmit_word`: the implementation will continue to handle
the original call and issue the originally scheduled callback
(as if the call that `Err`'d with `BUSY` never happened). However,
 it does not issue a callback for the call to `transmit_buffer`
  that returned `Err`.
```

I think this is what this is saying?

The valid error codes for `transmit_buffer` are:
  - `OFF`: the underlying hardware is not available, perhaps because it has
    not been initialized or has been initialized into a different mode
    (e.g., a USART has been configured to be a SPI).
  - `BUSY`: the UART is already transmitting and has not made a transmission

>     callback yet.

Reviewer: ```suggestion
    complete callback yet.
```

configured. The `Configure` trait allows a client to configure a UART,
by setting is baud date, data width, parity, stop bits, and whether
hardware flow control is enabled.

These two traits are separate because there are cases when clients
need to know the configuration but cannot set it. For example, when a UART
is virtualized across multiple clients (e.g., so multiple sources can
write to the console), individual clients may want to check the baud rate.
However, they cannot set the baud rate, because that is fixed and shared
across all of them. Similarly, some services may need to be able to set
the UART configuration but do not need to check it.

Most devices using serial ports today use 8-bit data, but some older
devices use more or fewer bits, and hardware implementations support
this. If the data width of a UART is set to less than 8 bits, data is
still partitioned into bytes, and the UART sends the least significant
bits of each byte. Suppose a UART is configured to send 7-bit
words. If a client sends 5 bytes, the UART will send 35 bits,
transmitting the bottom 7 bits of each byte. The most significant bit
of each byte is ignored.

> 

Reviewer: ```suggestion

To avoid ambiguity with in-memory bytes, we will refer to the 5- to 9-bit entities
sent on the wire as `characters`.
```

This change will require changing "data word" out in many places in this document,
but would bring the document in line with the terminology used in the Wikipedia article
linked at the beginning of the text as well as the datasheets I skimmed.

}

pub trait TransmitClient {
    fn transmitted_word(&self, _rval: Result<(), ErrorCode>) {}
    fn transmitted_buffer(
        &self,
        tx_buffer: &'static mut [u8],
        tx_len: usize,
        rval: Result<(), ErrorCode>,
    );
}
```

The `Transmit` trait has two data paths: `transmit_word` and
`transmit_buffer`.  The `transmit_word` method is used in narrow use
cases where the cost and complexity of buffer management is not
needed. Generally, software should use the `transmit_buffer`
method. Most software implementations use DMA, such that a call to
`transmit_buffer` triggers a single interrupt when the transfer
completes; this saves energy and CPU cycles over per-byte transfers
and also improves transfer speeds because hardware can keep the UART
busy.

Each byte transmitted is a data word for the UART. If the UART is
using 8-bit data words, each data word is a byte. If the UART is using
smaller data words, it MUST ignore the high order bits of the data
values. For example, if the UART is using 6-bit data words and is told
to transmit `0xff`, it will transmit `0x3f`, ignoring the first two
bits. If a client needs to transmit data words larger than 8 bits, it

> should use `transmit_word`, as `transmit_buffer` is a buffer of 8-bit

Reviewer: Do we want to limit `9N` people to no DMA? Maybe

> If a client needs to transmit characters larger than 8 bits, the byte array will be read as MSB halfwords. The unused upper bits MUST be set to `0`, that is for 9-bit characters: `buffer[0] ∈ [0,1]` and `buffer[1] ∈ [0-255]`.

bytes and cannot store 9-bit values.

There can be a single transmit operation ongoing at any
time. Successfully calling either `transmit_buffer` or
`transmit_word` causes the UART to become busy until it issues the
callback corresponding to the outstanding operation.

3.1 `transmit_buffer` and `transmitted_buffer`
===============================

`Transmit::transmit_buffer` sends a buffer of data. The result
returned by `transmit_buffer` indicates whether there will be a

> callback in the future. If `transmit_buffer` returns `Ok(())`,

Reviewer: ```suggestion
callback in the future. If `transmit_buffer` returns `Ok(())`, an
```

    not been initialized or has been initialized into a different mode
    (e.g., a USART has been configured to be a SPI).
  - `BUSY`: the UART is already transmitting and has not made a transmission
    complete callback yet.
  - `SIZE`: `tx_len` is larger than the passed slice.
  - `FAIL`: some other failure.

Calling `transmit_buffer` while there is an outstanding
`transmit_buffer` or `transmit_word` operation MUST return `Err(BUSY)`.

The `TransmitClient::transmitted_buffer` callback indicates completion
of a buffer transmission.  The `Result` indicates whether the buffer
was successfully transmitted.  The `tx_len` argument specifies how
many data words (defined by `Configure`) were transmitted. If the
`rval` of `transmitted_buffer` is `Ok(())`, `tx_len` MUST be equal to
the size of the transmission started by `transmit_buffer`, defined
above.  A call to `transmit_word` or `transmit_buffer` made within
this callback MUST NOT return `Err(BUSY)` unless it is because this is
not the first call to one of these methods in the callback.  When this
callback is made, the UART MUST be ready to receive another call. The
valid `ErrorCode` values for `transmitted_buffer` are all of those

> returned by `transmit_buffer` plus:

Reviewer: Do all of those errors make sense for the already-started callback case?
 - `OFF`: I can see this if some higher priority task pre-empted your HW access
 - `BUSY`: Same priority thought?
 - `SIZE`: I don't think this error should be possible given the narrow definition above (the slice length is known at call-time)
 - `FAIL`: Sure.

So maybe "except `SIZE`"?

  - `CANCEL` if the call to `transmit_buffer` was cancelled by a call
    to `abort` and the entire buffer was not transmitted.

>   - `SIZE` if the buffer could only be partially transmitted.

Reviewer: Why should this be `SIZE` instead of one of the other errors (or `FAIL`)?

3.2 `transmit_word` and `transmitted_word`
===============================

The `transmit_word` method transmits a single data word of data
asynchronously.  The word length is determined by the UART
configuration.  A UART implementation MAY choose to not implement
`transmit_word` and `transmitted_word`.  There is a default
implementation of `transmitted_word` so clients that do not use
`receive_word` do not have to implement a callback.

If `transmit_word` returns `Ok(())`, the implementation MUST call the
`transmitted_word` callback in the future. If a call to
`transmit_word` returns `Err`, the implementation MUST NOT issue a

> callback for this call, although if the it is `Err(BUSY)` is will

Reviewer: ```suggestion
callback for this call, although if the it is `Err(BUSY)` it will
```

issue a callback for the outstanding operation.  Valid `ErrorCode`
results for `transmit_word` are:
  - `OFF`: The underlying hardware is not available, perhaps because
    it has not been initialized or in the case of a shared
    hardware USART controller because it is set up for SPI.
  - `BUSY`: the UART is already transmitting and has not made a
    transmission callback yet.

>   - `NOSUPPORT`: the implementation does not support `transmit_word`
>     operations.

Reviewer: ```suggestion
```

I'm not sure we should allow this. If a chip really can't do this for some reason, then asking the implementing driver to statically allocate 1 word of memory and transparently convert this to a `transmit_buffer` internally is not a huge lift. I feel like runtime feature-failures like this should be avoided if we can.

Author: Take a look at how many implement it. ;)

  - `Err(()):` there was no outstanding request and there will be no future
    callback.

If there is no outstanding call to `transmit_word` or
`transmit_buffer`, `transmit_abort` MUST return `Err(())`.

4 `Receive` and `ReceiveClient` traits
===============================

The `Receive` and `ReceiveClient` traits are used to receive data from the
UART. They support both single-word and buffer reception. Buffer-based
reception is more efficient, as it allows an MCU to handle only one
interrupt for many data words. However, buffer-based reception only supports
data words of 6, 7, and 8 bits, so clients using 9-bit words need to use
word operations.

Each byte received is a data word for the UART. If the UART is using
8-bit data words, each data word is a byte. If the UART is using
smaller data words, it MUST zero the high order bits of the data
values. For example, if the UART is using 6-bit data words and
receives `0x1f`, it must store `0x1f` in a byte and not set high order
bits.  If the UART is using 9-bit words and receives `0x1ea`, it
stores this in a 32-bit value for `receive_word` as `0x000001ea`.

`Receive` supports a single outstanding receive request. A successful

> call to `receive_buffer` or `receive_word` causes the UART to be busy
> until the callback for the outstanding operation is issued.

Reviewer: ```suggestion
call to `receive_buffer` or `receive_word` causes the receive half of UART
to be busy until the callback for the outstanding operation is issued.
```

`received_word`.  There is a default implementation of `received_word`
so clients that do not use `receive_word` do not have to implement a
callback.

If the UART returns `Ok(())` to a call to `receive_word`, it MUST make
a `received_word` callback in the future, when it receives a data word
or some error occurs. Valid `Err` values of `receive_word` are:
  - `BUSY`: the UART is busy with an outstanding call to
    `receive_buffer` or `receive_word`.
  - `OFF`: the UART is powered down or in a configuration that does
    not allow UART reception (e.g., it is a USART in SPI mode).
  - `NOSUPPORT`: `receive_word` operations are not supported.
  - `FAIL`: some other error.

5 Composite Traits
===============================

In addition to the 6 basic traits, the UART HIL defines several traits
that use these basic traits as supertraits. These composite traits allow
structures to refer to multiple pieces of UART functionality with a
single reference and ensure that their implementations are coupled.

```rust
pub trait Uart<'a>: Configure + Configuration + Transmit<'a> + Receive<'a> {}
pub trait UartData<'a>: Configuration + Transmit<'a> + Receive<'a> {}

> pub trait UartAdvanced<'a>: Configure + Configuration + Transmit<'a> + ReceiveAdvanced<'a> {}

Reviewer: ```suggestion
```

We got rid of this, right?

    system call API.
  * `capsules::virtual_uart` provides a set of abstractions for virtualizing
    a single UART into many UARTs.

The structures in `capsules::virtual_uart` allow multiple clients to
read from and write to a serial port. Write operations are interleaved
at the granularity of `transmit_buffer` calls: each client's
`transmit_buffer` call is printed contiguously, but consecutive calls
to `transmit_buffer` from a single client may have other data inserted
between them. When a client calls `receive_buffer`, it starts reading
data from the serial port at that point in time, for the length of its
request. If multiple clients make `receive_buffer` calls that overlap
with one another, they each receive copies of the received data.

Suppose, for example, that there are two clients. One of them calls
`receive_buffer` for 8 bytes. A user starts typing "1234567890"
at the console. After the third byte, another client calls `receive_buffer`
for 4 bytes. After the user types "7", the second client will
receive a `received_buffer` callback with a buffer containing "4567".
After the user types "8", the first client will receive a callback

> with a buffer containing "12345678".

Reviewer: ```suggestion
with a buffer containing "12345678". If the second client then calls
`receive_buffer` for 1 byte, they will receive "9" (note that it never
sees "8", since that has been consumed by the time this second
receive call was made).
```

I think this is true?

`received_character`.  There is a default implementation of `received_character`
so clients that do not use `receive_character` do not have to implement a
callback.

If the UART returns `Ok(())` to a call to `receive_character`, it MUST make
a `received_character` callback in the future, when it receives a character
or some error occurs. Valid `Err` values of `receive_character` are:
  - `BUSY`: the UART is busy with an outstanding call to
    `receive_buffer` or `receive_character`.
  - `OFF`: the UART is powered down or in a configuration that does
    not allow UART reception (e.g., it is a USART in SPI mode).
  - `NOSUPPORT`: `receive_character` operations are not supported.
  - `FAIL`: some other error.

5 Composite Traits
===============================

In addition to the 6 basic traits, the UART HIL defines several traits
that use these basic traits as supertraits. These composite traits allow
structures to refer to multiple pieces of UART functionality with a
single reference and ensure that their implementations are coupled.

```rust
pub trait Uart<'a>: Configure + Configuration + Transmit<'a> + Receive<'a> {}

> pub trait UartData<'a>: Configuration + Transmit<'a> + Receive<'a> {}

Reviewer 2: ```suggestion
pub trait UartData<'a>:  Transmit<'a> + Receive<'a> {}
```
I thought the point of `UartData` was this it did not require `Configuration`?

Author: Good catch. Fixed.
